torch/csrc/jit/codegen/cuda/lower_expr_sort.cpp

```diff
@@ -0,0 +1,366 @@
+  auto seg_begin = exprs.begin();
+  auto seg_end = exprs.begin();
+  ScoreT seg_score = scores.at(*seg_begin);
+  while (seg_end != exprs.end()) {
+    const auto expr = *seg_end;
+    const auto cur_score = scores.at(expr);
+    if (seg_score == cur_score) {
+      // advance further
+      ++seg_end;
+      continue;
+    } else if (seg_score < cur_score) {
+      // segment ended
+      reorderSegmentBreadthFirst(seg_begin, seg_end);
+      seg_begin = seg_end;
+      seg_score = cur_score;
+    } else {
+      // exprs list is assumed to be sorted in the order of scores, so
+      // this should never be reachable
+      TORCH_INTERNAL_ASSERT(
+          false, "Unexpected expression: ", expr, ", score: ", cur_score);
+    }
+  }
+  reorderSegmentBreadthFirst(seg_begin, seg_end);
+}
+
+void mergeNonRootGroupsIntoRootGroups(
+    TargetGroupMapT& computed_at_exprs,
+    ExprTargetMapT& target_map) {
+  for (auto it = computed_at_exprs.begin(); it != computed_at_exprs.end();) {
+    TensorView* target = it->first;
+    if (target->hasComputeAt()) {
+      Expr* target_expr = target->definition();
+      TensorView* target_of_target = target_map.at(target_expr);
+      auto& target_group = computed_at_exprs.at(target_of_target);
+      auto pos =
+          std::find(target_group.begin(), target_group.end(), target_expr);
+      TORCH_INTERNAL_ASSERT(pos != target_group.end());
+      target_group.insert(pos, it->second.begin(), it->second.end());
+      // Update the target map
+      for (auto& inserted_expr : it->second) {
+        TORCH_INTERNAL_ASSERT(target_map.at(inserted_expr) == target);
+        target_map.at(inserted_expr) = target_of_target;
+      }
+      it = computed_at_exprs.erase(it);
+    } else {
+      ++it;
+    }
+  }
+}
+
+// Merge root loop-nests into reordered_exprs
+void mergeGroupsIntoSortedList(
+    TargetGroupMapT& computed_at_exprs,
+    ExprListT& reordered_exprs) {
+  while (computed_at_exprs.size() > 0) {
+    // Find the root loop-nest that has no dependency with the other
+    // loop-nests
+    TensorView* cur_target = computed_at_exprs.begin()->first;
+    for (auto& group : computed_at_exprs) {
+      auto target = group.first;
+      if (cur_target == target)
+        continue;
+      if (DependencyCheck::isDependencyOf(target, cur_target)) {
+        cur_target = target;
+      }
+    }
+    // cur_target can be visited
+    reordered_exprs.insert(
+        reordered_exprs.end(),
+        computed_at_exprs.at(cur_target).begin(),
+        computed_at_exprs.at(cur_target).end());
+    computed_at_exprs.erase(cur_target);
+  }
+}
+
+} // namespace
+
+// Reorder exprs so that LoopNestGenerator::handle(Expr*) can generate
+// correct loop nests. Vector exprs is assumed to be topologically
+// sorted, but that is not sufficient as tensors computed at
+// outer loops need to be located earlier.
+std::vector<Expr*> reorderExprsForComputeAt(const std::vector<Expr*>& exprs) {
```

Comment:
how do you feel about adding a FUSER_PERF_SCOPE here?

Reply:
I'm not sure whether this place would need its own perf scope. I think we could just add it later if it was found to be a performance concern.

Reply:
The intention for tracing is to be proactive: have the information available before you need it (ex. if you're investigating a perf issue, ideally you want the information w/o requiring a code modification + rebuild). I anticipate this will help most "in the field", where rebuilding is not easy. Trace scopes also allow a better understanding of the code paths.

Because of this I generally prefer to default to add scopes where we do non-trivial processing. 

That being said, I'll defer to you and @csarofeen to pick the right instrumentation scopes.

Reply:
I don't disagree with you in general, but at this point, I don't have a good idea of which processing would be non-trivial, maybe except for a few cases such as nvrtc compilation. So, it isn't easy for me to make an informed decision about where to put scope marking. I could just aggressively create scopes, but I wonder if that could result in a lot of noises when we actually needed to examine traces. 

Manual tracing may be incomplete, so using it for code understanding could be misleading.